### ISSUE TITLE
CI: Links in docs are not tested

### ISSUE THREAD

**Issue by errordeveloper** [cilium/cilium#10601]
I have attempted fixing some of the most trivial issues with links in our docs (https://github.com/cilium/cilium/pull/10576), but some links are not pass the link checker right now and I'm not sure how to fix these.
Perhaps we could instruct the checker to ignore these somehow? Or explore using an alternative link checker.

Here are some of the main examples of what's broken:

Anchors work fine in the browser, but not in the link checker:

```
(line   35) broken    https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html#cli-quick-configuration - Anchor 'cli-quick-configuration' not found
(line   19) broken    https://github.com/kubernetes-sigs/kubespray#requirements - Anchor 'requirements' not found
```

Commands that contain URL substrings:

```
(line   32) broken    https://github.com/weaveworks/eksctl/releases/download/latest_release/eksctl_$(uname - 404 Client Error: Not Found for url: https://github.com/weaveworks/eksctl/releases/download/latest_release/eksctl_$(uname
(line   39) broken    https://github.com/kubernetes/kops/releases/download/$(curl - 404 Client Error: Not Found for url: https://github.com/kubernetes/kops/releases/download/$(curl
```

This one needs auth:
```
(line   41) broken    https://github.com/cilium/cilium/projects/new - 404 Client Error: Not Found for url: https://github.com/cilium/cilium/projects/new
```

We also have a few localhost links like this:
```
(line   42) broken    http://localhost:9080 - HTTPConnectionPool(host='localhost', port=9080): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f31cddb0d10>: Failed to establish a new connection: [Errno 111] Connection refused'))
(line  238) broken    http://127.0.0.1:15000/logging?level=debug - HTTPConnectionPool(host='127.0.0.1', port=15000): Max retries exceeded with url: /logging?level=debug (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f31ce248bd0>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

**Comment by qmonnet** [comment]
[The linkcheck builder](http://www.sphinx-doc.org/en/master/_modules/sphinx/builders/linkcheck.html) can take a set of options:

``` python
    app.add_config_value('linkcheck_ignore', [], None)
    app.add_config_value('linkcheck_auth', [], None)
    app.add_config_value('linkcheck_retries', 1, None)
    app.add_config_value('linkcheck_timeout', None, None, [int])
    app.add_config_value('linkcheck_workers', 5, None)
    app.add_config_value('linkcheck_anchors', True, None)
    # Anchors starting with ! are ignored since they are
    # commonly used for dynamic pages
    app.add_config_value('linkcheck_anchors_ignore', ["^!"], None)
```

Do you think we could use these? Local links could be marked as ignored, and semi-broken anchors too (or we can just ignore them all with `linkcheck_anchors` set to `False` it seems).

We could possibly ignore links created with `curl`/`uname`, as I don't see an easy way to check the final target. I suppose ignoring links requiring auth is also the easiest, I don't see us putting GitHub credentials in the Makefile.

<details><summary>Something like that</summary>

``` diff
diff --git a/Documentation/conf.py b/Documentation/conf.py
index bf76ab2eec43..9b1ff2bff53a 100644
--- a/Documentation/conf.py
+++ b/Documentation/conf.py
@@ -230,6 +230,36 @@ http_strict_mode = False
 # Try as hard as possible to find references
 default_role = 'any'
 
+linkcheck_ignore = [
+    # Local links
+    'http://127.0.0.1',
+    'http://localhost',
+
+    # Incomplete URLs
+    'https:/$',
+    'https?://.*\$\((curl|uname)$',
+
+    # Needs authentication
+    'https://github.com/cilium/cilium/projects/new',
+    'https://github.com/cilium/cilium/settings/branches',
+    'https://jenkins.cilium.io/job/bisect-cilium/',
+    'https://jenkins.cilium.io/job/bisect-cilium/configure',
+    'https://jenkins.cilium.io/job/cilium-ginkgo/configure',
+
+    # Others
+    'http://elrepo.org/tiki/tiki-index.php', # Site filters on user-agent?
+    'https://api.twitter.com', # Broken but only used in example as blocked URL
+]
+ 
+linkcheck_anchors_ignore = [
+    'L[0-9]*-L[0-9]*$',
+    'getting-started-for-developers$',
+    'if-you-are-running-on-kubernetes$',
+    'layer-7$',
+    'requirements$',
+    'slide=id.g',
+    'stable-releases$',
+]
 
 def setup(app):
     app.add_stylesheet('parsed-literal.css')
```
</details>

On a side node, enabling link verification by default makes docs rendering way longer (it takes ~4 minutes on my setup).

**Comment by qmonnet** [comment]
We currently have a checker at: https://github.com/cilium/cilium/blob/191b672af20feda6e43f52fa1c429f1639cb50fd/Documentation/check-build.sh#L153
but it's commented out because of the issues reported here. First step would be to uncomment it, try to run locally, and look at what we get. Then fix or switch to something better if necessary/available.

Cc @zacharysarah for your information.

**Comment by vipul-21** [comment]
@qmonnet Can I pick up this issue and help fix the broken links ?

**Comment by qmonnet** [comment]
Sure thing, go ahead, thanks! There are two different things here, finding/fixing the broken links, and enabling a link checker (including running it in CI).

I haven't looked at this in a while, so not sure what the current state is regarding the links, or the commented checker.

Feel free to ping me on Slack or better yet, to open a thread in #sig-docs if necessary

**Comment by vipul-21** [comment]
Based on the `check_build.sh` there are three types of issue in the documentation currently:
- [ ] Links that are broken/not found
- [ ] `myst` not able to parse the `#anchor` in the Documentation. example: ```'myst' reference target not found```
- [ ] Circular inclusion in the `include directive`  example: ``` WARNING: circular inclusion in "include" directive: ```  
- [ ] Apart from the above issues, we need to enable the link checker for CI

**Comment by vipul-21** [comment]
@qmonnet I added the PR for broken links. For the second error I mentioned `'myst' reference target not found:`  does not seem to be an actual broken link but an error on `myst`. Wondering if we can upgrade the versions of `myst` and the other dependencies and check if the issue still exists. ( if so, can you help me on how I can do that ?)

Also, I see this function: https://github.com/cilium/cilium/blob/main/Documentation/check-build.sh#L36-L42 to ignore the myst warnings, should I use that ?

**Comment by qmonnet** [comment]
> For the second error I mentioned `'myst' reference target not found:` does not seem to be an actual broken link but an error on `myst`.

Sorry I don't follow, where did you see this error?

> Wondering if we can upgrade the versions of `myst` and the other dependencies and check if the issue still exists. ( if so, can you help me on how I can do that ?)

We can look into that if necessary. Let's sync on Slack sometime next week?

**Comment by vipul-21** [comment]
> Sorry I don't follow, where did you see this error?

When I use `make html` or `check_build.sh`

> We can look into that if necessary. Let's sync on Slack sometime next week?
Sure, let me try if I can fix it. If not will ping you on slack early  next week.

**Comment by qmonnet** [comment]
https://github.com/cilium/cilium/pull/26880 fixes the links but does not test them in CI (this will likely be https://github.com/cilium/cilium/pull/27116), let's keep this open.

**Comment by aanm** [comment]
@qmonnet given that https://github.com/cilium/cilium/pull/27116 was merged, can we close this GH issue?

**Comment by qmonnet** [comment]
@aanm I don't think so, we have a script to check the links, but we're [still not running it in CI](https://github.com/cilium/cilium/pull/27116#issuecomment-1752760611).
